pyfr/nputil.py

```diff
@@ -126,3 +127,187 @@ def npdtype_to_ctypestype(dtype):
         return None
 
     return _ctypestype_map[np.dtype(dtype).type]
+
+
+#Get indices for interval segments of interpolation
+def get_interval_idxs(t, xarr):
+    if np.asarray(xarr).size > 1:
+        idxs = np.array([np.max(np.nonzero(xx >= t[:-1] - 1e-12)[0]) 
+                                                        for xx in xarr])
+    else:
+        idxs = np.max(np.nonzero(xarr >= t[:-1] - 1e-12)[0])
+    return idxs
+
+#Modified Gauss Elimination/Thompson algorithm for tridiag systems
+#All changes are done in place and the algorithm 
+#is applied to the lhs and rhs together
+def trislv(a, b, c, d):
+    #forward elimination
+    d[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i] = bi1 = bi - ai*cdivb
+        #rhs     
+        d[i] -= ai*dj
+        d[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    for i, (ci, dj) in enumerate(zip(c[::-1], d[:0:-1]), start=1):
+        dd[i] -= ci*dj
+
+
+#tridiagonal solver for cyclic/periodic systems 
+#Using Sherman-Morrison formula and Thompson algorithm
+# A = A + u ⊗ v
+# u, v are vectors
+def trislvcyc(a, b, c, d, alpha, beta):
```

Comment:
It's been a while, but you should be able to have this call the regular `trislv` internally.

Reply:
I tried it but It is complicated and this is faster. I can do two calls to trislv but then it is not as faster as solving both systems in one set of loops.

Reply:
Your system is A*x = b = (A' + u*v^T)*x = b where A' is tridiagonal and u and v are vectors whose outer product gives the periodic components.  Whence solve A'*[y z ] = [u b] which is a tridiagonal system and should be *one* call to `trislv` with a stacked RHS.  Then x = z - (y*v^T*z)/(1 + v^T*y).  Entire function should only be four to five lines of code with roughly similar performance  to what your hand unrolled version delivers.  Moreover, this approach also enables the cyclic routine to automatically benefit from any fixes/improvements in `trislv`. 

As a general point any more complex solution to a problem must justify is existence by either providing additional functionality to the user, or demonstrably improved performance for the user.  Given this function is not in the critical path it is highly unlikely that even a five-fold increase in runtime would be noticeable.  Thus go with the simpler solution.

Reply:
Tried it before and I know it can be done by stacking but I just tried to use the same in-place concept. I do not think you still can keep trislv with in-place but I will retry again. Otherwise, I can modify trislv not to be inplace. Let me know if you have an idea to do it in-place or I can just modify trislv not to be in-place. I am open to suggestions here.

Reply:
You can still have it work in place easily enough.  Something like
```
def trislvcyc(a, b, c, d, alpha, beta):
    ud = np.hstack([d, np.zeros((len(d), 1))])
    ud[[0, -1], -1] = -b[0], alpha
    trislv(a, b, c, ud)
    
    d[:] = ud[:, :-1] - ...
```

Reply:
done